pkg/cmd/experimental/buildchain/buildchain.go

```diff
@@ -0,0 +1,367 @@
+				case "":
+					repo = join(cfg.Namespace, tr.ImageChange.From.Name)
+				default:
+					repo = join(tr.ImageChange.From.Namespace, tr.ImageChange.From.Name)
+				}
+
+				var tag string
+				switch tr.ImageChange.Tag {
+				case "":
+					tag = "latest"
+				default:
+					tag = tr.ImageChange.Tag
+				}
+
+				uniqueTag := true
+				for _, prev := range avoidDuplicates[repo] {
+					if prev == tag {
+						uniqueTag = false
+						break
+					}
+				}
+				if uniqueTag {
+					avoidDuplicates[repo] = append(avoidDuplicates[repo], tag)
+				}
+			}
+		}
+	}
+
+	return avoidDuplicates
+}
+
+// findRepoDeps accepts an image repository and a list of build
+// configurations and returns the dependency tree of the specified
+// image repository
+func findRepoDeps(repo, tag string, all bool, buildConfigList []buildapi.BuildConfig) (*ImageRepo, error) {
+	root := &ImageRepo{
+		FullName: repo,
+		Tags:     []string{tag},
+	}
+
+	namespace, name, err := split(repo)
+	if err != nil {
+		return nil, err
+	}
+
+	// Search all build configurations in order to find the image
+	// repositories depending on the specified image repository
+	var childNamespace, childName, childTag string
+	for _, cfg := range buildConfigList {
+		for _, tr := range cfg.Triggers {
+			// Setup zeroed fields to their default values
+			if tr.ImageChange != nil && tr.ImageChange.From.Namespace == "" {
+				tr.ImageChange.From.Namespace = cfg.Namespace
+			}
+			if tr.ImageChange != nil && tr.ImageChange.Tag == "" {
+				tr.ImageChange.Tag = "latest"
+			}
+
+			if tr.ImageChange != nil &&
+				namespace == tr.ImageChange.From.Namespace &&
+				name == tr.ImageChange.From.Name &&
+				tag == tr.ImageChange.Tag {
+
+				// Either To & Tag or DockerImageReference will be used as output
+				if cfg.Parameters.Output.To != nil && cfg.Parameters.Output.To.Name != "" {
+					childName = cfg.Parameters.Output.To.Name
+					childTag = cfg.Parameters.Output.Tag
+					if cfg.Parameters.Output.To.Namespace != "" {
+						childNamespace = cfg.Parameters.Output.To.Namespace
+					} else {
+						childNamespace = cfg.Namespace
+					}
+				} else {
+					ref, err := imageapi.ParseDockerImageReference(cfg.Parameters.Output.DockerImageReference)
+					if err != nil {
+						return nil, err
+					}
+					childName = ref.Name
+					childTag = ref.Tag
+					childNamespace = cfg.Namespace
+				}
+
+				childRepo := join(childNamespace, childName)
+
+				// Build all children and their dependency trees recursively
+				child, err := findRepoDeps(childRepo, childTag, all, buildConfigList)
+				if err != nil {
+					return nil, err
+				}
+
+				// Add edge between root and child
+				cfgFullName := join(cfg.Namespace, cfg.Name)
+				root.Edges = append(root.Edges, NewEdge(cfgFullName, child.FullName))
+
+				// If the user specifies the --all flag and the child depends
+				// on root via more than one tag, multiple instances of the child
+				// will make it into root.Children
+				if all {
```

Comment:
@bparees it's possible that two buildConfigs will have the same From.ICT.Tag (and same repo ofc) but two different Output.Tags of another single repository right? If so, I think this all check is wrong and this should be checked every time.